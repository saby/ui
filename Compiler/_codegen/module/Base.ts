/**
 */

import {
    IDependenciesController,
    createDependenciesController,
} from './Dependencies';
import { IExportsController, createExportsController } from './Exports';

/**
 * Интерфейс процессора, который выполняет компиляцию модуля в различный формат.
 */
export interface IModuleProcessor {
    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void;

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void;

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void;

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void;

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void;

    /**
     * Выполнить компиляцию модуля.
     * @returns {string} JavaScript код модуля.
     */
    compile(): string;
}

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 */
export abstract class Base implements IModuleProcessor {
    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Массив кода, который составляет тело будущего модуля.
     * @protected
     */
    protected code: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     */
    protected constructor() {
        this.name = '';
        this.useStrict = false;
        this.code = [];
        this.dependenciesController = createDependenciesController();
        this.exportsController = createExportsController();
    }

    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.code.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
