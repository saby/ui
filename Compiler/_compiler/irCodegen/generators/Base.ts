/**
 * @author Krylov M.A.
 *
 * Базовый генератор обращений к именованным сущностям.
 */

import type { TPrimitive, IFormatter } from '../Interface';
import type { TAliasesMap } from '../Aliases';

import { wrapSequence } from '../types/String';

export default class Base {
    protected constructor(
        protected readonly formatter: IFormatter,
        protected readonly instance: string,
        protected readonly aliases: TAliasesMap,
        protected readonly isRelease: boolean
    ) { }

    protected getPropertyName(fn: string): string {
        if (this.isRelease) {
            if (this.aliases.has(fn)) {
                return `.${this.aliases.get(fn)}/*${fn}*/`;
            }
        }

        return `.${fn}`;
    }

    protected generateMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${wrapSequence(args)})`;
    }

    protected generateFormattedMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${this.formatter.formatSequence(args, 1)})`;
    }
}
