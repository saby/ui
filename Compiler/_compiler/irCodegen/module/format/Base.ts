/**
 * @author Krylov M.A.
 */

import type { IModuleGenerator, IDependenciesController, IExportsController } from '../Interface';

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 *
 * @private
 */
export default abstract class Base implements IModuleGenerator {

    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Содержимое модуля.
     * @protected
     */
    protected contents: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     * @param {IDependenciesController} dependencies Контроллер зависимостей.
     * @param {IExportsController} exports Контроллер экспортируемых данных.
     * @protected
     */
    constructor(dependencies: IDependenciesController, exports: IExportsController) {
        this.name = '';
        this.useStrict = false;
        this.contents = [];

        this.dependenciesController = dependencies;
        this.exportsController = exports;
    }

    /**
     * Установить имя компилируемого модуля.
     * @param {string} name Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param {boolean} value Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param {string} ref Путь зависимости.
     * @param {string?} name Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param {string} block Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.contents.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param {string} block Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
