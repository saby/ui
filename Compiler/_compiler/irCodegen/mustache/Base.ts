/**
 * @author Krylov M.A.
 *
 * Базовый класс обработчиков mustache выражений.
 * Реализует основные операции над узлами.
 */

import type {
    Node,
    IExpressionVisitor,
    ArrayExpressionNode,
    BinaryExpressionNode,
    CallExpressionNode,
    ConditionalExpressionNode,
    DecoratorCallNode,
    DecoratorChainCallNode,
    DecoratorChainContext,
    EmptyStatementNode,
    ExpressionBrace,
    ExpressionStatementNode,
    IdentifierNode,
    LogicalExpressionNode,
    MemberExpressionNode,
    ObjectExpressionNode,
    ProgramNode,
    SequenceExpressionNode,
    ThisExpressionNode,
    UnaryExpressionNode,
    IObjectProperty
} from '../../expressions/Nodes';

import type {
    IMethods,
    IECMAScriptGenerator,
    ISymbols
} from '../Interface';

import type {
    IFlags,
    ISource,
    IMustacheMeta,
    IMustacheGenerator,
    IMustacheOptions
} from './Interface';

import {
    LiteralNode
} from '../../expressions/Nodes';

import Flags from './Flags';

import {
    wrapSequence,
    wrapArray,
    wrapString
} from '../types/String';

import { MUSTACHE_EXPRESSION_PARAMETERS } from '../Constants';

const WRAPPED_EMPTY_STRING = wrapString('');

/**
 * Create function parameters for function that represents mustache expression.
 * @param {Flags} flags Content dependent flags for mustache expression.
 */
function getFunctionParams(flags: IFlags): string[] {
    const params = [...MUSTACHE_EXPRESSION_PARAMETERS];

    if (flags.hasChildrenReference) {
        return params;
    }

    // no children variable usage. remove it from parameters
    params.pop();
    if (flags.hasContextReference) {
        return params;
    }

    // no context variable usage. remove it from parameters
    params.pop();
    if (flags.hasFuncContextReference) {
        return params;
    }

    // no funcContext variable usage. remove it from parameters
    params.pop();
    if (flags.hasMethodsReference) {
        return params;
    }

    // perhaps, function uses rk function or this variable
    return [];
}

export abstract class Generator<TContext extends IMustacheOptions>
    implements IExpressionVisitor<TContext, string>, IMustacheGenerator<string>
{
    protected readonly methods: IMethods<string>;
    protected readonly generator: IECMAScriptGenerator;
    protected readonly symbols: ISymbols;
    protected readonly source: ISource;

    protected flags: Flags;
    protected prevFlags: Flags[];

    /**
     * Initialize new instance of mustache expression generator.
     * @param {IMethods<string>} methods Methods generator.
     * @param {IECMAScriptGenerator} generator ECMAScript generator.
     * @param {ISymbols} symbols In-module string literals accessor generator.
     * @param {ISource} source Source identifiers allowed to use inside mustache expression.
     * @protected
     */
    constructor(
        methods: IMethods<string>,
        generator: IECMAScriptGenerator,
        symbols: ISymbols,
        source: ISource
    ) {
        this.methods = methods;
        this.generator = generator;
        this.symbols = symbols;
        this.source = source;

        this.flags = new Flags();
        this.prevFlags = [];
    }

    abstract visitDecoratorChainCallNode(node: DecoratorChainCallNode, context: TContext): string;

    abstract visitDecoratorChainContext(node: DecoratorChainContext, context: TContext): string;

    abstract visitDecoratorCallNode(node: DecoratorCallNode, context: TContext): string;

    /**
     * Generate meta information for mustache expression.
     * @param {Node} node Node instance.
     * @param {<TContext>} options Code generation options.
     */
    generate(node: Node, options: TContext): IMustacheMeta<string> {
        this.flags.reset();

        if (options.alwaysTableFunction) {
            this.flags.alwaysTableFunction = true;
        }

        options.extraParameters = options.extraParameters ?? [];

        const body = node.accept(this, options);
        const isTableFunction = this.flags.isTableFunction();

        return {
            isTableFunction,
            flags: this.flags,
            body: isTableFunction ? this.generateFunction(body, this.flags, options) : body,
            program: node.string,
            shouldEscape: options.shouldEscape
        };
    }

    visitProgramNode(node: ProgramNode, context: TContext): string {
        // Не поддерживается наличие в body более чем 1 выражения. Нужен вывод предупреждения?
        // Больше одного выражения - это когда они разделены ';'.
        const body = node.body[0].accept(this, context);

        return this.wrapProgramBody(body, node, context);
    }

    visitEmptyStatementNode(node: EmptyStatementNode, context: TContext): string {
        return '';
    }

    visitExpressionStatementNode(node: ExpressionStatementNode, context: TContext): string {
        return node.expression.accept(this, context);
    }

    visitThisExpressionNode(node: ThisExpressionNode, context: TContext): string {
        this.flags.hasSelfReference = true;

        return 'this';
    }

    visitArrayExpressionNode(node: ArrayExpressionNode, context: TContext): string {
        const elements = node.elements.map((element: Node) => element.accept(this, context));

        return wrapArray(elements);
    }

    visitObjectExpressionNode(node: ObjectExpressionNode, context: TContext): string {
        const pairs = [];

        // eslint-disable-next-line guard-for-in
        for (const property in node.properties) {
            const key = this.visitObjectPropertyName(node.properties[property].key, context);

            if (key) {
                const value = node.properties[property].value.accept(this, context);

                if (context.generateObjectPropertyAsIdentifier !== false) {
                    if (key === value) {
                        pairs.push(` ${key}`);
                        continue;
                    }
                }

                pairs.push(` ${key}: ${value}`);
            }
        }

        return `{${wrapSequence(pairs, ',')} }`;
    }

    visitSequenceExpressionNode(node: SequenceExpressionNode, context: TContext): string {
        return wrapSequence(
            node.expressions
                .map((expression: Node) => expression.accept(this, context))
        );
    }

    visitUnaryExpressionNode(node: UnaryExpressionNode, context: TContext): string {
        return node.operator + node.argument.accept(this, context);
    }

    visitBinaryExpressionNode(node: BinaryExpressionNode, context: TContext): string {
        let operandsContext = context;

        if (node.operator === '+') {
            // Принудительно приводим потенциально отсутствующий alternate при конкатенации
            operandsContext = {
                ...context,
                defaultAlternateValue: WRAPPED_EMPTY_STRING
            };
        }

        const left = node.left.accept(this, operandsContext);
        const right = node.right.accept(this, operandsContext);

        return `${left} ${node.operator} ${right}`;
    }

    visitLogicalExpressionNode(node: LogicalExpressionNode, context: TContext): string {
        const left = node.left.accept(this, context);
        const right = node.right.accept(this, context);

        return `${left} ${node.operator} ${right}`;
    }

    visitConditionalExpressionNode(node: ConditionalExpressionNode, context: TContext): string {
        const test = node.test.accept(this, context);
        const consequent = node.consequent.accept(this, context);

        if (typeof node.alternate !== 'undefined') {
            const alternate = node.alternate.accept(this, context);

            return `(${test} ? ${consequent} : ${alternate})`;
        }

        return `(${test} ? ${consequent} : ${context.defaultAlternateValue})`;
    }

    visitCallExpressionNode(node: CallExpressionNode, context: TContext): string {
        const callee = node.callee.accept(this, context);
        const args = node.arguments.map(arg => arg.accept(this, context));

        return `${callee}(${wrapSequence(args)})`;
    }

    visitMemberExpressionNode(node: MemberExpressionNode, context: TContext): string {
        const object = node.object.accept(this, context);
        const property = node.property.accept(this, context);

        if (node.computed) {
            return `${object}[${property}]`;
        }

        return `${object}.${property}`;
    }

    visitIdentifierNode(node: IdentifierNode, context: TContext): string {
        return node.name;
    }

    visitExpressionBrace(node: ExpressionBrace, context: TContext): string {
        return `(${node.name.accept(this, context)})`;
    }

    visitLiteralNode(node: LiteralNode, context: TContext): string {
        if (typeof node.value === 'string') {
            const uLiteral = context.shouldUnescape ? unescape(node.value) : node.value;

            if (context.allowAllocatingLiterals) {
                return this.symbols.access(uLiteral);
            }

            return wrapString(uLiteral);
        }

        return `${node.value}`;
    }

    protected visitObjectPropertyName(key: IObjectProperty['key'], context: TContext): string {
        if (key instanceof LiteralNode) {
            return key.accept(this, {
                ...context,
                allowAllocatingLiterals: false
            });
        }

        if (context.generateObjectPropertyAsIdentifier !== false) {
            return key.name;
        }

        return this.symbols.access(key.name);
    }

    /**
     * Wrap mustache expression body with required function provided with context.
     * @param {string} body Mustache expression body.
     * @param {ProgramNode} node Program node of mustache expression.
     * @param {<TContext>} context Current generating context.
     * @protected
     */
    protected wrapProgramBody(body: string, node: ProgramNode, context: TContext): string {
        let result = body;

        if (context.shouldWrapUndefined === true) {
            this.flags.hasMethodsReference = true;

            result = this.methods.wrapUndefined(body);
        }

        if (context.shouldSanitize === true) {
            this.flags.hasMethodsReference = true;

            return this.methods.sanitize(result);
        }

        return result;
    }

    /**
     * Generate function for mustache expression.
     * @param {Flags} flags Content dependent flags.
     * @param {string} body Expression content.
     * @param {<TContext>} context Current generating context.
     * @protected
     */
    protected generateFunction(body: string, flags: IFlags, context: TContext): string {
        if (context.allowReducingFunctionParameters !== false && context.extraParameters?.length > 0) {
            throw new Error(
                'внутренняя ошибка генерации кода: extraParameters запрещено использовать вместе с allowReducingFunctionParameters'
            );
        }

        const params = context.allowReducingFunctionParameters !== false
            ? getFunctionParams(flags)
            : [...MUSTACHE_EXPRESSION_PARAMETERS, ...context.extraParameters];

        if (flags.hasSelfReference) {
            // Function contains usage of this variable. Arrow functions are forbidden!

            return this.generator.toAnonymousFunction(`return ${body};`, params);
        }

        return this.generator.toArrowExpression(body, params);
    }

    /**
     * Generate mustache expression meta information of particular sub node.
     * @param {Node} node Particular node of expression.
     * @param {<TContext>} context Current generating context.
     * @protected
     */
    protected generateMustacheMeta(node: Node, context: TContext): IMustacheMeta<string> {
        this.enterExpression(new Flags());
        const body = node.accept(this, context);
        const flags = this.leaveExpression();

        return {
            body,
            flags,
            program: node.string,
            isTableFunction: flags.isTableFunction(),
            shouldEscape: context.shouldEscape === true
        };
    }

    /**
     * Start processing sub expression which is expected to exported avoiding entry function
     * and switch processing context.
     * @param {Flags} flags Sub-expression flags.
     * @protected
     */
    protected enterExpression(flags: Flags): void {
        this.prevFlags.push(this.flags);

        this.flags = flags;
    }

    /**
     * Finish processing sub expression end restore processing context to its previous state.
     * @protected
     */
    protected leaveExpression(): Flags {
        const prev = this.flags;

        this.flags = this.prevFlags.pop();
        this.flags.merge(prev);

        return prev;
    }
}
