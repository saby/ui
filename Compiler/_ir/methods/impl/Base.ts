/**
 * @author Krylov M.A.
 */

import type { IGenerator as IMarkupGenerator, IGeneratorConfig } from 'UICommon/Executor';

import type { IPrivateMethods } from '../Interface';
import type { IAttributes, IDataArray, IteratorScope } from '../Interface';

/**
 * Базовый класс, предоставляющий общие методы для набора методов markup и internal.
 * Реализует методы интерфейса IMinifiedMethods.
 *
 * @private
 */
export default abstract class Base implements IPrivateMethods {

//# region Lazy getters

    private lazyViewDecorators: {
        [name: string]: Function;
    };

    protected get viewDecorators() {
        if (!this.lazyViewDecorators) {
            this.lazyViewDecorators = require('View/decorators');
        }

        return this.lazyViewDecorators;
    }

//# endregion

//# region implementation of interface IMinifiedMethods

    t(content: unknown): unknown {
        return this.sanitize(content);
    }

    u(value: { _moduleName: string; }): unknown {
        return this.wrapUndefined(value);
    }

    S(value: unknown): string {
        return this.wrapString(value);
    }

    g(dataSource: unknown, path: string[]): unknown {
        return this.getter(dataSource, path);
    }

    s(dataSource: unknown, path: string[], value: unknown): unknown {
        return this.setter(dataSource, path, value);
    }

    d(name: string, args?: unknown[]): unknown {
        return this.decorate(name, args);
    }

    c(funcContext: unknown, dataSource: unknown, path: string[], args?: unknown[]): unknown {
        return this.call(funcContext, dataSource, path, args);
    }

    C(dataSource: unknown, path: string[], args?: unknown[]): unknown {
        return this.call2(dataSource, path, args);
    }

    r(url: string): string {
        return this.getResourceURL(url);
    }

    D(data: unknown): unknown {
        return this.dots(data);
    }

//# endregion

//# region interface IMethods

    abstract filterOptions(options: unknown): unknown;

    abstract sanitize(content: unknown): unknown;

    abstract wrapUndefined(value: { _moduleName: string; }): unknown;

    abstract wrapString(value: unknown): string;

    abstract getResourceURL(url: string): string;

    abstract getter(
        dataSource: unknown,
        path: string[]
    ): unknown;

    abstract setter(
        dataSource: unknown,
        path: string[],
        value: unknown
    ): unknown;

    abstract decorate(
        name: string,
        args?: unknown[]
    ): unknown;

    abstract call(
        funcContext: unknown,
        dataSource: unknown,
        path: string[],
        args?: unknown[]
    ): unknown;

    abstract call2(
        dataSource: unknown,
        path: string[],
        args?: unknown[]
    ): unknown;

    abstract dots(data: unknown): unknown;

//# endregion

//# region interface IPrivateMethods

    abstract validateNodeKey(key: number | string): number | string;

    abstract createScope(scope: object): Record<string, unknown>;

    abstract getContext(obj: unknown): unknown;

    abstract getTypeFunc(
        name: string,
        data: unknown
    ): unknown;

    abstract createDataArray(
        array: unknown[],
        fileName: string,
        isWasabyTemplate: boolean,
        isVdom: boolean
    ): IDataArray;

    abstract uniteScope(
        inner: unknown,
        outer: unknown
    ): Function;

    abstract plainMerge(
        inner: unknown,
        outer: unknown,
        cloneFirst?: unknown
    ): unknown;

    abstract processMergeAttributes(
        inner: IAttributes,
        outer: IAttributes
    ): IAttributes;

    abstract plainMergeAttr(
        inner: unknown,
        outer: unknown
    ): unknown;

    abstract plainMergeContext(
        inner: unknown,
        outer: unknown
    ): unknown;

    abstract calcParent(
        obj: unknown,
        propertyName: string,
        data: unknown
    ): unknown;

    abstract presetScope(
        object: unknown,
        data: unknown,
        key: unknown,
        iteratorScope: IteratorScope
    ): unknown;

    abstract createGenerator(
        isVdom: boolean,
        forceCompatible: boolean,
        config: IGeneratorConfig
    ): IMarkupGenerator;

    abstract templateError(
        fileName: string,
        error: Error,
        data: unknown
    ): void;

    abstract isolateScope(
        scope: object,
        data: object,
        propertyName?: string
    ): Record<string, unknown>;

//# endregion
}
