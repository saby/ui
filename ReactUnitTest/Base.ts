import { Control } from 'UICore/Base';
import { IControlOptions } from 'UICommon/Base';

export interface ITestOptions extends IControlOptions {
    testName?: string;
    afterMountCallback?: Function;
    afterUpdateCallback?: Function;
}

export class TestBaseControl<
    IOptions extends ITestOptions = ITestOptions
> extends Control<IOptions> {
    testName: string = '';
    afterMountCallback: Function = null;
    afterUpdateCallback: Function = null;
    _beforeMount(options: ITestOptions): void {
        this.afterMountCallback = options.afterMountCallback;
        this.afterUpdateCallback = options.afterUpdateCallback;
        this.testName = options.testName;
    }
    _afterMount(): void {
        this.afterMountCallback?.(this);
    }
    _afterUpdate(): void {
        this.afterUpdateCallback?.(this);
    }
}
