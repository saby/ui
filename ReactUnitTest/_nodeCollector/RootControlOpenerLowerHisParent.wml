<div class="{{ className }}">
   <ReactUnitTest._nodeCollector.ControlWithOpener setOpener="{{ setOpener }}" opener="{{ savedOpener }}" />
</div>
