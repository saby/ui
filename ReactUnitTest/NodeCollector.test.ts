/**
 * @jest-environment jsdom
 */
import { act } from 'react-dom/test-utils';
import { unmountComponentAtNode } from 'react-dom';

import { Control } from 'UICore/Base';
import { goUpByControlTree } from 'UICore/NodeCollector';

import RootControlOpenerLowerHisParent from './_nodeCollector/RootControlOpenerLowerHisParent';

describe('NodeCollector', () => {
    let container: HTMLElement;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        jest.useFakeTimers();
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(setTimeout);
    });

    afterEach(() => {
        jest.useRealTimers();
        jest.restoreAllMocks();
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    describe('goUpByControlTree', () => {
        test('При зацикливании выдаётся ошибка, без бесконечной рекурсии', () => {
            const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
            act(() => {
                Control.createControl(RootControlOpenerLowerHisParent, {}, container);
            });
            // Ждём все таймеры: Сначала маунт, потом из _afterMount новая перерисовка с присваиванием плохого опенера.
            jest.runAllTimers();

            expect(container).toMatchSnapshot('Убедимся, что дождались: класс должен поменяться');

            let hasUncaughtException = false;
            try {
                goUpByControlTree(document.getElementById('startNodeCollectorHere'));
            } catch (e) {
                hasUncaughtException = true;
            }

            expect(hasUncaughtException).toBeFalsy();
            expect(consoleErrorSpy.mock.calls.length).toBe(1);

            const message = consoleErrorSpy.mock.calls[0][1];
            expect(message).toContain(
                'ReactUnitTest/_nodeCollector/LowerOpener -> ReactUnitTest/_nodeCollector/ControlWithOpener -> ReactUnitTest/_nodeCollector/LowerOpener'
            );
        });
    });
});
