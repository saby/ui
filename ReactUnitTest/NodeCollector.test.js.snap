// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeCollector goUpByControlTree При зацикливании выдаётся ошибка, без бесконечной рекурсии: Убедимся, что дождались: класс должен поменяться 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div
    class="setOpenerClassName"
  >
    <div
      class="someClass"
      tabindex="0"
    >
      <div
        class="someClass"
        tabindex="0"
      >
        <span
          id="startNodeCollectorHere"
        >
          some text
        </span>
      </div>
    </div>
  </div>
</div>
`;
